sweep/bucket.go

```diff
@@ -0,0 +1,186 @@
+package sweep
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/lnwallet"
+)
+
+const (
+	// DefaultMinFeeRate is the default minimum fee rate we'll allow when
+	// sweeping inputs. This value corresponds to a 1 sat/vbyte fee rate.
+	DefaultMinFeeRate = lnwallet.FeePerKwFloor
+
+	// DefaultMaxFeeRate is the default maximum fee rate we'll allow when
+	// sweeping inputs. This value corresponds to a 1e4 sat/vbyte fee rate.
+	DefaultMaxFeeRate = DefaultMinFeeRate + (DefaultMinFeeRate * 1e4)
+
+	// DefaultFeeRateSpacing is the default distribution of each fee rate
+	// bucket. The current value allows for a distribution as follows:
+	//
+	//   bucket #1: min=1 sat/vbyte, max=20 sat/vbyte
+	//   bucket #2: min=21 sat/vbyte, max=40 sat/vbyte...
+	DefaultFeeRateSpacing = DefaultMinFeeRate * 20
```

Comment:
I think 20 might be too large a distance between buckets. Especially at the current fee levels on the chain today. Perhaps 5 is more reasonable?

Reply:
Suppose the fee landscape changes drastically, are we going to make a new release that changes this constant?

I am still not convinced that the bucket system is better than just grouping inputs based on their exact fee rate. Inputs that are offered with the same conf target will end up in the same bucket anyway and for bumping non-lnd inputs (not the main use case as far as I can see), there needs to be some awareness with the caller that grouping only happens with the same fee rate. A UI could offer a limited set of options to make that clear.

Reply:
Agree 20 is probably too large. Ended up going with 10.

> Suppose the fee landscape changes drastically, are we going to make a new release that changes this constant?

Reply:
> I am still not convinced that the bucket system is better than just grouping inputs based on their exact fee rate.

It's a tradeoff between 1 transaction per satoshi fee rate unit compared to grouping them optimistically to batch the transactions. The former approach will result in more transaction on chain, and likely amount of satoshis spent on fees by `lnd` which may offset any precision gains. The buckets are about "satisficing": the caller specifies a fee rate and we do our best to meet that match, but if we can bundle them, then we do as it leads to less fees globally across all callers.

Reply:
We can add a strict mode in the future if there're certain cases that are unable to compromise w.r.t their target fee rate.

Reply:
Conf based sweeps will end up in the same group anyway, because they have the exact same fee rate. And for inputs specified on the command line, I think a user would have a few favorite fee rates that they pick from. I don't see it happening that a user bumps fees for several inputs in the same time window with just slightly different fee rates. So I don't expect there to be more chain txes in real life use of this feature.

But similar to the design discussion, this is also based on an expectation of how the function will be used.

Reply:
Regardless of how it ends up getting used (user access patterns), it's a fact that more chain transactions will be generated if no sort of bucketization is used. As an example, today we use 3 or 6 confs pretty widely within the codebase. With the current proposed buckets those would be grouped together meaning a single transaction rather than two.

Reply:
Why would they be grouped together? It could be that the fee rates for 3 and 6 confs from the estimator at some point end up in a different bucket?

Reply:
Ah sorry, there's an implicit assumption there for the example that those would fall in similar buckets. Main point is that we can create less on chain transactions by satisficing across fee rates.

Reply:
Yes I got that. My point is that if we only ever pick a few fee rates (for example 3 confs, 6 confs (for internal sweeps) and 10,20,50 sat/byte (for external bumps)), we don't need to cluster and can just publish a tx per distinct fee rate.

Reply:
Yes if there's really no variation in fee preferences we'll end up publishing more transactions, but may save on fee cost globally.